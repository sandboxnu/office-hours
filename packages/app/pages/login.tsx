import Router from "next/router";
import { ReactElement, useEffect, useState } from "react";
import { LockOutlined, UserOutlined } from "@ant-design/icons";
import { message, Button, Form, Input, Select, Spin, Modal } from "antd";
//import Select from "react-select";
import styled from "styled-components";
import { API } from "@koh/api-client";

const Container = styled.div`
  margin-left: auto;
  margin-right: auto;
  text-align: center;
  width: 300px;
  padding-top: 100px;
`;
class option {
  value: number;
  label: string;
}
export default function Login(): ReactElement {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [loginInfo, setLoginInfo] = useState(null);
  const [cOptions, setCOptions] = useState<option[]>(null);
  const [selectedCourse, setSelectedCourse] = useState<number>(null);
  useEffect(() => {
    getCourses();
    if (cOptions) {
      setSelectedCourse(cOptions[0].value);
    }
  }, []);
  const getCourses = async () => {
    await API.course.getAllCourses().then((courses) => {
      if (!courses) {
        message.warning("courses not found");
        return;
      }
      const CourseOptions = courses.map((course) => ({
        value: course.id,
        label: course.name,
      }));
      setCOptions(CourseOptions);
    });
  };
  function onChange(value: number) {
    setSelectedCourse(value);
  }
  function submit(values: any) {
    // get all course ids with userid
    // if courseids.length==1 log in using login(values, course), if not stop. >2 then set modal to true.
    // modal calls modalSubmit that sends info to login to login.
    setLoginInfo(values);
    const courses = {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        email: values.username,
        password: values.password,
      }),
    };
    fetch(`/api/v1/getAllcourses`, courses)
      .then(async (response) => {
        const data = await response.json();
        if (!response.ok) {
          // get error message from body or default to response statusText
          const error = (data && data.message) || response.statusText;
          if (data.message === "Invalid credential") {
            message.error("Invalid password.");
          } else if (data.message === "NotInCourse") {
            message.error("Not registered in selected course");
          } else {
            message.error("User Not Found");
          }
          return Promise.reject(error);
        } else {
          if (data.length == 1) {
            login(data[0].courseId);
          } else if (data.length > 1) {
            setIsModalOpen(true);
          }
        }
      })
      .catch((error) => {
        console.error("There was an error!", error);
      });
  }
  const handleOk = () => {
    setIsModalOpen(false);
    login(selectedCourse);
  };
  const handleCancel = () => {
    setIsModalOpen(false);
  };
  function login(course: number) {
    const loginRequest = {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        email: loginInfo.username,
        password: loginInfo.password,
      }),
    };
    fetch(`/api/v1/ubc_login/${course}`, loginRequest)
      .then(async (response) => {
        const data = await response.json();
        if (!response.ok) {
          // get error message from body or default to response statusText
          const error = (data && data.message) || response.statusText;
          if (data.message === "Invalid credential") {
            message.error("Invalid password.");
          } else if (data.message === "NotInCourse") {
            message.error("Not registered in selected course");
          } else {
            message.error("User Not Found");
          }
          return Promise.reject(error);
        } else {
          Router.push(`/api/v1/login/entry/${course}?token=${data.token}`);
        }
      })
      .catch((error) => {
        console.error("There was an error!", error);
      });
  }
  //put token inside login request
  if (!cOptions) {
    return <Spin tip="Loading..." size="large" />;
  } else {
    return (
      <Container>
        <Form
          name="normal_login"
          className="login-form"
          initialValues={{ remember: true }}
          onFinish={submit}
        >
          <h1>UBC HelpMe</h1>
          <Form.Item
            name="username"
            rules={[{ required: true, message: "Please input your Username!" }]}
          >
            <Input
              prefix={<UserOutlined className="site-form-item-icon" />}
              placeholder="Username"
            />
          </Form.Item>
          <Form.Item
            name="password"
            rules={[{ required: true, message: "Please input your Password!" }]}
          >
            <Input
              prefix={<LockOutlined className="site-form-item-icon" />}
              type="password"
              placeholder="Password"
            />
          </Form.Item>

          <Form.Item>
            <a
              style={{ float: "right", marginTop: "-10px" }}
              href="./forgetpassword/forget"
            >
              Forgot password
            </a>
          </Form.Item>

          <Form.Item style={{ marginTop: "-15px" }}>
            <Button
              style={{ width: "100%", marginTop: "-15px" }}
              type="primary"
              htmlType="submit"
              className="login-form-button"
            >
              Log in
            </Button>
            Or <a href="./signup/signup">register now!</a>
          </Form.Item>
        </Form>
        <Modal
          title="Basic Modal"
          visible={isModalOpen}
          onOk={handleOk}
          onCancel={handleCancel}
        >
          <Select
            placeholder="Select course"
            onChange={onChange}
            options={cOptions}
            defaultValue={selectedCourse}
          />
        </Modal>
      </Container>
    );
  }
}
