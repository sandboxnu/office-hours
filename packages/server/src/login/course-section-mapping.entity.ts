import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  ManyToOne,
  JoinColumn,
} from 'typeorm';
import { CourseModel } from '../course/course.entity';

@Entity('course_section_mapping_model')
export class CourseSectionMappingModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ nullable: true })
  crn: number;

  // Represents the course that this maps to
  @ManyToOne((type) => CourseModel)
  @JoinColumn({ name: 'courseId' })
  course: CourseModel;

  @Column({ nullable: true })
  courseId: number;
}
