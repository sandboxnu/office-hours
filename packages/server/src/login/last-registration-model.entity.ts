import { UserModel } from '../profile/user.entity';
import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  JoinColumn,
  OneToOne,
} from 'typeorm';
import { Season } from '@koh/common';

@Entity('last_registration_model')
export class LastRegistrationModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @OneToOne((type) => UserModel)
  @JoinColumn({ name: 'profId' })
  prof: UserModel;

  @Column()
  profId: number;

  @Column('text')
  lastRegisteredSemester: string; // same format as what's passed by Admin
}

export const khourySemesterCodes: Record<Season, string> = {
  Fall: '10',
  Spring: '30',
  Summer_1: '40',
  Summer_Full: '50',
  Summer_2: '60',
};
