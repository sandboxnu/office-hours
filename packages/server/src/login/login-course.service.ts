import {
  isKhouryCourse,
  KhouryDataParams,
  KhouryProfCourse,
  Role,
  Season,
} from '@koh/common';
import { Injectable } from '@nestjs/common';
import { CourseModel } from 'course/course.entity';
import { CourseSectionMappingModel } from 'login/course-section-mapping.entity';
import { UserCourseModel } from 'profile/user-course.entity';
import { UserModel } from 'profile/user.entity';
import { SemesterModel } from 'semester/semester.entity';
import { Connection } from 'typeorm';
import { ProfSectionGroupsModel } from './prof-section-groups.entity';
import { khourySemesterCodes } from './last-registration-model.entity';

@Injectable()
export class LoginCourseService {
  constructor(private connection: Connection) {}

  public async addUserFromKhoury(info: KhouryDataParams): Promise<UserModel> {
    let user: UserModel;
    const neuEmail = info.email.replace('@husky.neu.edu', '@northeastern.edu');
    user = await UserModel.findOne({
      where: { email: neuEmail },
      relations: ['courses', 'courses.course'],
    });

    if (!user) {
      user = await UserModel.create({
        courses: [],
        email: neuEmail,
        firstName: info.first_name,
        lastName: info.last_name,
        hideInsights: [],
      }).save();
    }

    const userCourses = [];

    for (const c of info.courses) {
      if (isKhouryCourse(c)) {
        const course = await this.courseCRNToCourse(c.crn, c.semester);

        if (course) {
          const userCourse = await this.courseToUserCourse(
            user.id,
            course.id,
            this.convertKhouryRole(c.role),
          );
          userCourses.push(userCourse);
        }
      } else {
        // c is a KhouryProfCourse
        // only need to inspect one of the CRNs from the list
        // b/c they should all map to the same section group
        if (c.crns.length !== 0) {
          const courseCRN = c.crns[0];
          const profCourse = await this.courseCRNToCourse(
            courseCRN,
            c.semester,
          );

          if (profCourse) {
            const profUserCourse = await this.courseToUserCourse(
              user.id,
              profCourse.id,
              Role.PROFESSOR,
            );
            userCourses.push(profUserCourse);
          }
        }
      }
    }

    // Delete "stale" user courses
    for (const previousCourse of user.courses) {
      if (
        !this.hasUserCourse(userCourses, previousCourse) &&
        previousCourse.course.enabled
      ) {
        if (!previousCourse.override) {
          previousCourse.remove();
        } else {
          userCourses.push(previousCourse);
        }
      }
    }

    // If Prof, save the JSON data
    if (info.courses[0] && !isKhouryCourse(info.courses[0])) {
      const profSectionGroups = await ProfSectionGroupsModel.findOne({
        where: { profId: user.id },
      });
      if (profSectionGroups) {
        profSectionGroups.sectionGroups = info.courses as KhouryProfCourse[];
      } else {
        await ProfSectionGroupsModel.create({
          profId: user.id,
          sectionGroups: info.courses,
        }).save();
      }
    }

    user.courses = userCourses;
    await user.save();
    return user;
  }

  public async courseCRNToCourse(
    courseCRN: number,
    semester: string, // 6-digit semester code
  ): Promise<CourseModel> {
    const semModel = await this.getSemester(semester);
    const courseSectionModel =
      await CourseSectionMappingModel.createQueryBuilder('section_mapping')
        .leftJoinAndSelect('section_mapping.course', 'course')
        .where(
          'section_mapping.crn = :courseCRN and course.semesterId = :semesterId',
          { courseCRN, semesterId: semModel.id },
        )
        .getOne();

    return courseSectionModel?.course;
  }

  public async courseToUserCourse(
    userId: number,
    courseId: number,
    role: Role,
  ): Promise<UserCourseModel> {
    let userCourse: UserCourseModel;
    userCourse = await UserCourseModel.findOne({
      where: { userId, courseId },
    });
    if (userCourse && userCourse.override && userCourse.role === role) {
      userCourse.override = false;
      await userCourse.save();
    }
    if (!userCourse) {
      userCourse = await UserCourseModel.create({
        userId,
        courseId,
        role,
      }).save();
    }
    return userCourse;
  }

  // parses 6-digit semester code, where first 4 digits represent year and last 2 digits represent academic semester (ex: 202110)
  public parseKhourySemester(khourySemester: string): {
    season: Season;
    year: number;
  } {
    // parsing time
    let year = Number(khourySemester.slice(0, 4));
    const semesterCode = khourySemester.slice(-2);
    const season = Object.keys(khourySemesterCodes).find(
      (key) => khourySemesterCodes[key] === semesterCode,
    ) as Season;
    // edge case for Fall semester, included in the next academic year
    if (season === 'Fall') {
      year--;
    }

    return { season, year };
  }

  private async getSemester(khourySemester: string) {
    const { season, year } = this.parseKhourySemester(khourySemester);
    let semModel = await SemesterModel.findOne({ where: { season, year } });
    if (!semModel) {
      semModel = await SemesterModel.create({
        season,
        year,
        courses: [],
      }).save();
    }
    return semModel;
  }

  private hasUserCourse(
    userCourses: UserCourseModel[],
    previousCourse: UserCourseModel,
  ): boolean {
    return userCourses.some(
      (uc) =>
        uc.courseId === previousCourse.courseId &&
        uc.userId === previousCourse.userId &&
        uc.role === previousCourse.role,
    );
  }

  // util functions for converting Khoury Admin data to KOH data
  private convertKhouryRole(khouryRole: 'TA' | 'Student'): Role {
    return khouryRole.toLowerCase() === 'ta' ? Role.TA : Role.STUDENT;
  }
}
