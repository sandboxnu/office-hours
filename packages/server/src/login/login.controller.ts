import {
  ERROR_MESSAGES,
  GetSelfEnrollResponse,
  KhouryDataParams,
  KhouryRedirectResponse,
  Role,
  UBCOloginParam,
} from '@koh/common';
import {
  BadRequestException,
  Body,
  Controller,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Post,
  Query,
  Req,
  Res,
  UnauthorizedException,
  UseGuards,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { JwtService } from '@nestjs/jwt';
import * as Sentry from '@sentry/node';
import { CourseModel } from 'course/course.entity';
import { User } from 'decorators/user.decorator';
import { Request, Response } from 'express';
import { JwtAuthGuard } from 'guards/jwt-auth.guard';
//import * as httpSignature from 'http-signature';
import { UserCourseModel } from 'profile/user-course.entity';
import * as bcrypt from 'bcrypt';
import { UserModel } from 'profile/user.entity';
// import { questionEMail } from 'readline-sync';
import { Connection } from 'typeorm';
import { NonProductionGuard } from '../guards/non-production.guard';
import { LoginCourseService } from './login-course.service';

@Controller()
export class LoginController {
  constructor(
    private connection: Connection,
    private loginCourseService: LoginCourseService,
    private jwtService: JwtService,
    private configService: ConfigService,
  ) {}
  //front post to this
  @Post('/ubc_login')
  async receiveDataFromLogin(
    @Req() req: Request,
    @Res() res: Response,
    @Body() body: UBCOloginParam,
  ): Promise<any> {
    UserModel.findOne({
      where: { email: body.email },
    })
      .then(async (user) => {
        if (!user) {
          return res.status(404).send({ message: 'User Not found' });
        }

        const token = await this.jwtService.signAsync(
          { userId: user.id },
          { expiresIn: 60 },
        );
        if (token === null || token === undefined) {
          console.error('Temporary JWT is invalid');
          throw new HttpException(
            ERROR_MESSAGES.loginController.invalidTempJWTToken,
            HttpStatus.INTERNAL_SERVER_ERROR,
          );
        }
        bcrypt.compare(body.password, user.password, (err, data) => {
          //if error than throw error
          if (err) throw err;

          //if both match than you can do anything
          if (data) {
            return res.status(200).send({ token, ...body });
          } else {
            return res.status(401).json({ message: 'Invalid credential' });
          }
        });
      })
      .catch((err) => {
        res.status(500).send({ message: err.message });
      });
  }

  @Post('/khoury_login')
  async receiveDataFromKhoury(
    @Req() req: Request,
    @Body() body: KhouryDataParams,
  ): Promise<KhouryRedirectResponse> {
    let user;
    try {
      user = await this.loginCourseService.addUserFromKhoury(body);
    } catch (e) {
      Sentry.captureException(e);
      console.error('login threw an exception, the body was ', body);
      console.error(e);
      throw new HttpException(
        ERROR_MESSAGES.loginController.addUserFromKhoury,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
    // Create temporary login token to send user to.
    const token = await this.jwtService.signAsync(
      { userId: user.id },
      { expiresIn: 60 },
    );

    if (token === null || token === undefined) {
      console.error('Temporary JWT is invalid');
      throw new HttpException(
        ERROR_MESSAGES.loginController.invalidTempJWTToken,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }

    return {
      redirect:
        this.configService.get('DOMAIN') + `/api/v1/login/entry?token=${token}`,
    };
  }

  // NOTE: Although the two routes below are on the backend,
  // they are meant to be visited by the browser so a cookie can be set

  // This is the real admin entry point, Kevin changed to also just take a user id, change to that sign in only
  @Get('/login/entry')
  async enterUBCOH(
    @Res() res: Response,
    @Query('token') token: string,
  ): Promise<void> {
    const isVerified = await this.jwtService.verifyAsync(token);

    if (!isVerified) {
      throw new UnauthorizedException();
    }

    const payload = this.jwtService.decode(token) as { userId: number };

    // if (payload === null || payload === undefined) {
    //   console.error('Decoded JWT is invalid');
    //   throw new HttpException(
    //     ERROR_MESSAGES.loginController.invalidPayload,
    //     HttpStatus.INTERNAL_SERVER_ERROR,
    //   );
    // }

    this.enter(res, payload.userId);
  }

  // This is for login on development only
  @Get('/login/dev')
  @UseGuards(NonProductionGuard)
  async enterFromDev(
    @Res() res: Response,
    @Query('userId') userId: number,
  ): Promise<void> {
    this.enter(res, userId);
  }

  // Set cookie and redirect to proper page
  private async enter(res: Response, userId: number) {
    // Expires in 30 days
    const authToken = await this.jwtService.signAsync({
      userId,
      expiresIn: 60 * 60 * 24 * 30,
    });

    if (authToken === null || authToken === undefined) {
      console.error('Authroziation JWT is invalid');
      throw new HttpException(
        ERROR_MESSAGES.loginController.invalidTempJWTToken,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }

    const isSecure = this.configService
      .get<string>('DOMAIN')
      .startsWith('https://');
    res
      .cookie('auth_token', authToken, { httpOnly: true, secure: isSecure })
      .redirect(302, '/');
  }

  @Get('/logout')
  async logout(@Res() res: Response): Promise<void> {
    const isSecure = this.configService
      .get<string>('DOMAIN')
      .startsWith('https://');
    res
      .clearCookie('auth_token', { httpOnly: true, secure: isSecure })
      .redirect(302, '/login');
  }

  @Get('self_enroll_courses')
  async selfEnrollEnabledAnywhere(): Promise<GetSelfEnrollResponse> {
    const courses = await CourseModel.find();
    return { courses: courses.filter((course) => course.selfEnroll) };
  }

  @Post('create_self_enroll_override/:id')
  @UseGuards(JwtAuthGuard)
  async createSelfEnrollOverride(
    @Param('id') courseId: number,
    @User() user: UserModel,
  ): Promise<void> {
    const course = await CourseModel.findOne(courseId);

    if (!course.selfEnroll) {
      throw new UnauthorizedException(
        'Cannot self-enroll to this course currently',
      );
    }

    const prevUCM = await UserCourseModel.findOne({
      where: {
        courseId,
        userId: user.id,
      },
    });

    if (prevUCM) {
      throw new BadRequestException(
        'User already has an override for this course',
      );
    }

    await UserCourseModel.create({
      userId: user.id,
      courseId: courseId,
      role: Role.STUDENT,
      override: true,
      expires: true,
    }).save();
  }
}
