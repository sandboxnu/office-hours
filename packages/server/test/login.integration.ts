import { JwtService } from '@nestjs/jwt';
import { TestingModuleBuilder } from '@nestjs/testing';
import { LoginModule } from '../src/login/login.module';
import {
  OrganizationFactory,
  UserCourseFactory,
  UserFactory,
} from './util/factories';
import { setupIntegrationTest } from './util/testUtils';
import * as bcrypt from 'bcrypt';
import { OrganizationUserModel } from 'organization/organization-user.entity';

const mockJWT = {
  signAsync: async (payload) => JSON.stringify(payload),
  verifyAsync: async (payload) => JSON.parse(payload).token !== 'INVALID_TOKEN',
  decode: (payload) => JSON.parse(payload),
};

describe('Login Integration', () => {
  const supertest = setupIntegrationTest(
    LoginModule,
    (t: TestingModuleBuilder) =>
      t.overrideProvider(JwtService).useValue(mockJWT),
  );

  describe('POST /ubc_login', () => {
    it('returns 400 if no email is provided', async () => {
      await supertest()
        .post('/ubc_login')
        .send({ password: 'fake_password' })
        .expect(400);
    });

    it('returns 404 if user not found', async () => {
      await supertest()
        .post('/ubc_login')
        .send({ email: 'fake_email@ubc.ca', password: 'fake_password' })
        .expect(404);
    });

    it('returns 401 if password is incorrect', async () => {
      const user = await UserFactory.create({ password: 'real_password' });
      await mockJWT.signAsync({ userId: user.id });

      await supertest()
        .post('/ubc_login')
        .send({ email: user.email, password: 'invalid_password' })
        .expect(401);
    });

    it('returns 403 if account is deactivated', async () => {
      const salt = await bcrypt.genSalt(10);
      const password = await bcrypt.hash('realpassword', salt);

      const user = await UserFactory.create({
        password: password,
        accountDeactivated: true,
      });
      await mockJWT.signAsync({ userId: user.id });

      await supertest()
        .post('/ubc_login')
        .send({ email: user.email, password: 'realpassword' })
        .expect(403);
    });

    it('returns 200 if password is correct', async () => {
      const salt = await bcrypt.genSalt(10);
      const password = await bcrypt.hash('realpassword', salt);

      const user = await UserFactory.create({ password: password });
      await mockJWT.signAsync({ userId: user.id });

      const res = await supertest()
        .post('/ubc_login')
        .send({ email: user.email, password: 'realpassword' });

      expect(res.body).toMatchSnapshot();
      expect(res.status).toBe(200);
    });
  });

  describe('POST /login/entry', () => {
    it('request to entry with correct jwt payload works', async () => {
      const token = await mockJWT.signAsync({ userId: 1 });

      const res = await supertest()
        .get(`/login/entry?token=${token}`)
        .expect(302);

      expect(res.header['location']).toBe('/courses');
      expect(res.get('Set-Cookie')[0]).toContain('userId');
    });

    it('entry as user with courses goes to root page', async () => {
      const user = await UserFactory.create();
      await UserCourseFactory.create({ user: user });
      const token = await mockJWT.signAsync({ userId: user.id });

      const res = await supertest()
        .get(`/login/entry?token=${token}`)
        .expect(302);

      expect(res.header['location']).toBe('/courses');
      expect(res.get('Set-Cookie')[0]).toContain('userId');
    });

    it('request to entry with invalid jwt returns error', async () => {
      const token = await mockJWT.signAsync({ token: 'INVALID_TOKEN' });

      await supertest().get(`/login/entry?token=${token}`).expect(401);
    });
  });

  describe('POST /ubc_login', () => {
    it('returns 400 if no email is provided', async () => {
      await supertest()
        .post('/ubc_login')
        .send({ password: 'fake_password' })
        .expect(400);
    });

    it('returns 404 if user not found', async () => {
      await supertest()
        .post('/ubc_login')
        .send({ email: 'fake_email@ubc.ca', password: 'fake_password' })
        .expect(404);
    });

    it('returns 401 if password is incorrect', async () => {
      const user = await UserFactory.create({ password: 'real_password' });
      await mockJWT.signAsync({ userId: user.id });

      await supertest()
        .post('/ubc_login')
        .send({ email: user.email, password: 'invalid_password' })
        .expect(401);
    });

    it('returns 401 when organization does not allow legacy auth', async () => {
      const organization = await OrganizationFactory.create({
        legacyAuthEnabled: false,
      });
      const user = await UserFactory.create({ password: 'real_password' });

      await OrganizationUserModel.create({
        userId: user.id,
        organizationId: organization.id,
      }).save();

      await mockJWT.signAsync({ userId: user.id });

      await supertest()
        .post('/ubc_login')
        .send({ email: user.email, password: 'real_password' })
        .expect(401);
    });

    it('returns 401 when user did not sign up with legacy account system', async () => {
      const organization = await OrganizationFactory.create({
        legacyAuthEnabled: false,
      });
      const user = await UserFactory.create({ password: null });

      await OrganizationUserModel.create({
        userId: user.id,
        organizationId: organization.id,
      }).save();

      await mockJWT.signAsync({ userId: user.id });

      await supertest()
        .post('/ubc_login')
        .send({ email: user.email, password: 'real_password' })
        .expect(401);
    });

    it('returns 200 if password is correct', async () => {
      const salt = await bcrypt.genSalt(10);
      const password = await bcrypt.hash('realpassword', salt);

      const user = await UserFactory.create({ password: password });
      await mockJWT.signAsync({ userId: user.id });

      const res = await supertest()
        .post('/ubc_login')
        .send({ email: user.email, password: 'realpassword' });

      expect(res.body).toMatchSnapshot();
      expect(res.status).toBe(200);
    });
  });

  describe('GET /logout', () => {
    it('makes sure logout endpoint is destroying cookies like a mob boss', async () => {
      const res = await supertest().get(`/logout`).expect(302);
      expect(res.header['location']).toBe('/login');
      expect(res.get('Set-Cookie')[0]).toContain('auth_token=;');
    });
  });
});
