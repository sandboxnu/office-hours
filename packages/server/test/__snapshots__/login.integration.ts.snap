// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Integration POST /ubc_login returns 200 if password is correct 1`] = `
Object {
  "email": "user@ubc.ca",
  "password": "realpassword",
  "token": "{\\"userId\\":1}",
}
`;
